rlbot/src/state_builder.rs

```diff
@@ -0,0 +1,269 @@
+use rlbot_flat::flat::{DesiredBallState, DesiredCarState, DesiredGameState, DesiredMatchInfo, DesiredPhysics, Rotator, RotatorPartial, Vector3, Vector3Partial};
+
+/// Extension methods for easy construction of [DesiredGameState].
+///
+/// Example:
+/// ```rust
+/// use rlbot::state_builder::{DesiredCarStateExt, DesiredGameStateExt, DesiredPhysicsExt};
+/// use rlbot_flat::flat::{DesiredGameState, Vector3};
+/// let mut dgs = DesiredGameState::default();
+/// dgs.mod_car(0, |c| {
+///     c.set_location(Vector3::default());
+///     c.set_boost(100.);
+/// });
+/// dgs.mod_balls(0..5, |_, b| {
+///     b.set_location_z(0.);
+///     b.set_velocity_z(0.);
+/// });
+/// ```
+pub trait DesiredGameStateExt {
+
+    fn mod_match_info(
+        &mut self,
+        build: impl FnOnce(&mut DesiredMatchInfo),
+    );
+
+    fn mod_car(
+        &mut self,
+        index: usize,
+        build: impl FnOnce(&mut DesiredCarState),
+    );
+
+    fn mod_cars(
+        &mut self,
+        indices: impl IntoIterator<Item = usize>,
+        build: impl Fn(usize, &mut DesiredCarState),
+    );
```

Comment:
I still think I prefer the following, as its a bit more flexible
```suggestion
    fn mod_cars(
        &mut self,
        desired_car_states: impl IntoIterator<Item = (usize, DesiredCarState)>,
    );
```
It makes usage of the function a bit more clear on what it does imo:
```rust
dgs.mod_cars((0..5).map(|i| {
    let mut dcs = DesiredCarState::default();
    dcs.mod_physics(|p| p.set_location(Vector3::default()));
    (i, dcs)
}));
```
and also lets you do this
```rust
dgs.mod_cars([
    (0, DesiredCarState::default()),
    (2, my_car_state_from_somewhere_else),
]);
```

Reply:
Maybe this would be better so it matches mod_car 🤔 
```suggestion
    fn mod_cars(
        &mut self,
        desired_car_states: impl IntoIterator<Item = (usize, impl Fn(&mut DesiredCarState))>,
    );
```
Previous examples would look like:
```rust
dgs.mod_cars((0..5).map(|i| {
    (i, |dcs: &mut DesiredCarState| {
        dcs.mod_physics(|p| p.set_location(Vector3::default()))
    })
}));
```
and
```rust
dgs.mod_cars([
    (0, |_| ()),
    (2, |x: &mut DesiredCarState| {
        *x = my_car_state_from_somewhere_else
    }),
]);
```

Reply:
I could go with the last suggestion, but I do not see why anyone would use `mod_cars` and treat each car differently like your last example. They could just call `mod_car` twice instead.